Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+
+Expressions are mostly used to control the flow of the if statement.
+
+### Values
+
+A _Value_ is a single concrete form of one of the supported types.
+All expressions, subexpressions, variables and functions, must resolve down into a value before it can be used.
+Examples of values include `1`, `true`, `"General Kenobi"`.
+
+### Types
+
+Yarn supports the following types and these must be supported by an implementing program:
+
+- number
+- boolean
+- string
+
+The precision, storage, and form of the number internally is unspecified, however it must support decimals.
+As an example of this in C# the `Decimal`, `Complex`, and `float` formats are valid (though some make more sense than others) but `int` is not.
+Numbers in expressions can be written as either integers or as decimals, but decimal numbers must use the `.` symbol as the decimal seperator of the the number.
+
+Strings must be capable of holding utf-8 values as this is what the yarn language is written in, but the internals of this is unspecified provided all valid utf-8 characters are supported.
+Strings in expressions must be encapsulated between `"` and `"` symbols.
+
+Booleans must be capable of representing the boolean logic values of `true` and `false`, however the specific implementation is undefined.
+Booleans must not be exposed as `1` and `0` to expressions even if they are represented this way internally by the implementing program.
+Booleans in expressions must be written as `true` for true and `false` for false.
+
+Additional types supported are unspecified.
+
+### Variables
+
+_Variables_ are a means of associating a value with a name so that it can be more easily used and changed in multiple places.
+Variables can only be used inside of expressions.
+
+#### Naming and scope
+
+All variables must start with the `$` symbol followed immediately by any combination of `a`-`z`, `A` - `Z`, and `0-9`.
+Arbitrary symbol variable names are allowed by wrapping the symbols after the `$` inside backtick symbols \`.
+Any symbols except another backtick are allowed in an arbitrary sybmol variable name.
+This is useful to allow writers to create semantic connections between variables.
+The minimum and maximum length of a variable name is unspecified.
+
+Variable names must be unique throughout their scope.
+The _scope_ of the variable is what defines for how long a variable exists.
+Once a variable leaves its scope it is no longer valid and implementing programs must not support accessing variables outside of their scope.
+The scope of variables is undefined, however implementing programs should err on the side of the smallest possible scope for a variable.
+
+`$name` is an example of a standard variable name, `$`\``character.name`\` is an example of an arbitrary sybmol variable name.
+
+#### Types
+
+Variables have a type which represents which of the supported type' values it can hold.
+Once a variable has its type determined either by declaration or inference it cannot change.
```

Comment:
How are variables declared, other than being `<<set>>`?

Reply:
Yarn Spinner 2.0 introduces a `<<declare $variable_name as TYPE>>` syntax (see discussion in https://github.com/YarnSpinnerTool/YarnSpinner/issues/260 for full information.) This will need to be added to the spec.